torchqdynamics/ode/odeint.py

```diff
@@ -0,0 +1,240 @@
+            solution at these time points.
+    """
+    tspan = init_tspan(tspan)
+    y0, tspan, solver = init_solver(y0, tspan, solver)
+    save_at = init_saveat(save_at, tspan, backward_mode)
+
+    # Dispatch to appropriate solver
+    if sensitivity in (None, "autograd"): #TODO: Separate these two cases
+        if solver.solver_type == "rk":
+            return odeint_adaptive(f, y0, tspan, solver, save_at, atol, rtol, backward_mode)
+        elif solver.solver_type == "out":
+            return odeint_outsource(f, y0, tspan, solver, save_at, backward_mode)
+    elif sensitivity == "adjoint":
+        return odeint_adjoint(f, y0, tspan, solver, save_at, atol, rtol)
+
+
+# -------------------------------------------------------------------------------------------------
+#     ODE subroutines
+# -------------------------------------------------------------------------------------------------
+
+def odeint_adjoint(f, y0, tspan, solver='dopri5', save_at=(), atol=1e-8, rtol=1e-6):
+    """Solve an ODE with an adjoint method in the backward pass."""
+    # Check inputs
+    if not isinstance(f, nn.Module) or count_params(f) == 0:
+        raise TypeError("For adjoint-based autodiff, `f` must be an instance of nn.Module "
+                        "with a non-empty set of parameters.")
+
+    if not (hasattr(f.__class__, 'forward_adj') and callable(getattr(f.__class__, 'forward_adj'))):
+        raise TypeError("For adjoint-based autodiff, `f` must have a method "
+                        "`forward_adj(self, t, a)`.")
+
+    # Extract model parameters
+    params = torch.cat([p.view(-1) for p in f.parameters() if p.requires_grad])
+    f_adj = f.forward_adj
+    # Run the ODE integrator
+    t, y = ODEIntAdjoint.apply(f, f_adj, y0, tspan, params, solver, save_at, atol, rtol)
+    return t, y
+
+def odeint_adaptive(f_, y0, tspan, solver, save_at=(), atol=1e-8, rtol=1e-6, backward_mode=False):
+    """Core ODE solver subroutine for adaptive step size solvers. Solves an ODE of the form
+    `dy / dt = f(t, y(t))`."""
+    # Check for backward mode
+    if backward_mode:
+        f = lambda t, y: -f_(-t, y)
+        tspan = -tspan.flip(0)
+        save_at = -save_at.flip(0)
+    else: f = f_
+
+    # Initialize save arrays
+    t0, T = tspan[0], tspan[-1]
+    t_saved = torch.zeros((len(save_at),), dtype=t0.dtype)
+    y_saved = torch.zeros((len(save_at),) + y0.shape, dtype=y0.dtype)
+    ckpt_counter, ckpt_flag = 0, False
+    if len(save_at) > 0 and save_at[0] == t0:
+        t_saved[0], y_saved[0] = t0, y0
+        ckpt_counter += 1
+
+    # Initialize the ODE routine
+    f0 = f(t0, y0)
+    dt = init_step(f, f0, y0, t0, solver.order, atol, rtol)
+    
+    # Run the ODE routine
+    t, y, ft = t0, y0, f0
+    while t < T:         
+        # Check if checkpointing required, or final time is reached
+        if len(save_at) > 0 and (ckpt_counter < len(save_at)) and (t + dt >= save_at[ckpt_counter]):
+            dt_old, ckpt_flag = dt, True
```

Comment:
I guess `ckpt` is `checkpoint` but maybe it's better to put explicit name for non common words